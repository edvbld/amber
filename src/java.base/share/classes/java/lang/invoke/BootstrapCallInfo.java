/*
 * Copyright (c) 2017, Oracle and/or its affiliates. All rights reserved.
 * DO NOT ALTER OR REMOVE COPYRIGHT NOTICES OR THIS FILE HEADER.
 *
 * This code is free software; you can redistribute it and/or modify it
 * under the terms of the GNU General Public License version 2 only, as
 * published by the Free Software Foundation.  Oracle designates this
 * particular file as subject to the "Classpath" exception as provided
 * by Oracle in the LICENSE file that accompanied this code.
 *
 * This code is distributed in the hope that it will be useful, but WITHOUT
 * ANY WARRANTY; without even the implied warranty of MERCHANTABILITY or
 * FITNESS FOR A PARTICULAR PURPOSE.  See the GNU General Public License
 * version 2 for more details (a copy is included in the LICENSE file that
 * accompanied this code).
 *
 * You should have received a copy of the GNU General Public License version
 * 2 along with this work; if not, write to the Free Software Foundation,
 * Inc., 51 Franklin St, Fifth Floor, Boston, MA 02110-1301 USA.
 *
 * Please contact Oracle, 500 Oracle Parkway, Redwood Shores, CA 94065 USA
 * or visit www.oracle.com if you need additional information or have any
 * questions.
 */

package java.lang.invoke;

import java.lang.invoke.MethodHandles.Lookup;
import java.util.Arrays;
import java.util.List;

/**
 * An interface providing full static information about a particular
 * call to a
 * <a href="package-summary.html#bsm">bootstrap method</a> of an
 * dynamic call site or dynamic constant.
 * This information includes the method itself, the associated
 * name and type, and any associated static arguments.
 * <p>
 * If a bootstrap method declares exactly two arguments, and is
 * not of variable arity, then it is fed only two arguments by
 * the JVM, the {@linkplain Lookup lookup object} and an instance
 * of {@code BootstrapCallInfo} which supplies the rest of the
 * information about the call.
 * <p>
 * The API for accessing the static arguments allows the bootstrap
 * method to reorder the resolution (in the constant pool) of the
 * static arguments, and to catch errors resulting from the resolution.
 * This mode of evaluation <em>pulls</em> bootstrap parameters from
 * the JVM under control of the bootstrap method, as opposed to
 * the JVM <em>pushing</em> parameters to a bootstrap method
 * by resolving them all before the bootstrap method is called.
 * @apiNote
 * <p>
 * The {@linkplain Lookup lookup object} is <em>not</em> included in this
 * bundle of information, so as not to obscure the access control
 * logic of the program.
 * In cases where there are many thousands of parameters, it may
 * be preferable to pull their resolved values, either singly or in
 * batches, rather than wait until all of them have been resolved
 * before a constant or call site can be used.
 * <p>
 * A push mode bootstrap method can be adapted to a pull mode
 * bootstrap method, and vice versa.  For example, this generic
 * adapter pops a push-mode bootstrap method from the beginning
 * of the static argument list, eagerly resolves all the remaining
 * static arguments, and invokes the popped method in push mode.
 * The callee has no way of telling that it was not called directly
 * from the JVM.
 * <blockquote><pre>{@code
static Object genericBSM(Lookup lookup, BootstrapCallInfo<Object> bsci)
    throws Throwable {
  ArrayList<Object> args = new ArrayList<>();
  args.add(lookup);
  args.add(bsci.invocationName());
  args.add(bsci.invocationType());
  MethodHandle bsm = (MethodHandle) bsci.get(0);
  List<Object> restOfArgs = bsci.asList().subList(1, bsci.size());
  // the next line eagerly resolves all remaining static arguments:
  args.addAll(restOfArgs);
  return bsm.invokeWithArguments(args);
}
 * }</pre></blockquote>
 *
 * <p>
 * In the other direction, here is a combinator which pops
 * a pull-mode bootstrap method from the beginning of a list of
 * static argument values (already resolved), reformats all of
 * the arguments into a pair of a lookup and a {@code BootstrapCallInfo},
 * and invokes the popped method.  Again the callee has no way of
 * telling it was not called directly by the JVM, except that
 * all of the constant values will appear as resolved.
 * Put another way, if any constant fails to resolve, the
 * callee will not be able to catch the resulting error,
 * since the error will be thrown by the JVM before the
 * bootstrap method is entered.
 * <blockquote><pre>{@code
static Object genericBSM(Lookup lookup, String name, Object type,
                         MethodHandle bsm, Object... args)
    throws Throwable {
  BootstrapCallInfo<Object> bsci = makeBootstrapCallInfo(bsm, name, type, args);
  return bsm.invoke(lookup, bsci);
}
 * }</pre></blockquote>
 *
 * @since 11
 */
public
interface BootstrapCallInfo<T> extends ConstantGroup {

    /// Access

    /**
     * Returns the number of static arguments.
     * @return the number of static arguments
     */
    int size();

    /**
     * Returns the selected static argument, resolving it if necessary.
     * Throws a linkage error if resolution proves impossible.
     * @param index which static argument to select
     * @return the selected static argument
     * @throws LinkageError if the selected static argument needs resolution
     * and cannot be resolved
     */
    Object get(int index) throws LinkageError;

    /**
     * Returns the selected static argument,
     * or the given sentinel value if there is none available.
     * If the static argument cannot be resolved, the sentinel will be returned.
     * If the static argument can (perhaps) be resolved, but has not yet been resolved,
     * then the sentinel <em>may</em> be returned, at the implementation's discretion.
     * To force resolution (and a possible exception), call {@link #get(int)}.
     * @param index the selected constant
     * @param ifNotPresent the sentinel value to return if the static argument is not present
     * @return the selected static argument, if available, else the sentinel value
     */
    Object get(int index, Object ifNotPresent);

    /**
     * Returns an indication of whether a static argument may be available.
     * If it returns {@code true}, it will always return true in the future,
     * and a call to {@link #get(int)} will never throw an exception.
     * <p>
     * After a normal return from {@link #get(int)} or a present
     * value is reported from {@link #get(int,Object)}, this method
     * must always return true.
     * <p>
     * If this method returns {@code false}, nothing in particular
     * can be inferred, since the query only concerns the internal
     * logic of the {@code BootstrapCallInfo} object which ensures that a
     * successful query to a constant will always remain successful.
     * The only way to force a permanent decision about whether
     * a static argument is available is to call {@link #get(int)} and
     * be ready for an exception if the constant is unavailable.
     * @param index the selected constant
     * @return {@code true} if the selected static argument is known by
     *     this object to be present, {@code false} if it is known
     *     not to be present or
     */
    boolean isPresent(int index);


    /// Views

    /**
     * Create a view on the static arguments as a {@link List} view.
     * Any request for a static argument through this view will
     * force resolution.
     * @return a {@code List} view on the static arguments which will force resolution
     */
    default List<Object> asList() {
        return new AbstractConstantGroup.AsList(this, 0, size());
    }

    /**
     * Create a view on the static argument as a {@link List} view.
     * Any request for a static argument through this view will
     * return the given sentinel value, if the corresponding
     * call to {@link #get(int,Object)} would do so.
     * @param ifNotPresent the sentinel value to return if a static argument is not present
     * @return a {@code List} view on the static arguments which will not force resolution
     */
    default List<Object> asList(Object ifNotPresent) {
        return new AbstractConstantGroup.AsList(this, 0, size(), ifNotPresent);
    }


    /// Bulk operations

    /**
     * Copy a sequence of static arguments into a given buffer.
     * This is equivalent to {@code end-offset} separate calls to {@code get},
     * for each index in the range from {@code offset} up to but not including {@code end}.
     * For the first static argument that cannot be resolved,
     * a {@code LinkageError} is thrown, but only after
     * preceding static arguments have been stored.
     * @param start index of first static argument to retrieve
     * @param end limiting index of static arguments to retrieve
     * @param buf array to receive the requested static arguments
     * @param pos position in the array to offset storing the static arguments
     * @return the limiting index, {@code end}
     * @throws LinkageError if a static argument cannot be resolved
     */
    default int copyArguments(int start, int end,
                              Object[] buf, int pos)
            throws LinkageError
    {
        int bufBase = pos - start;  // buf[bufBase + i] = get(i)
        for (int i = start; i < end; i++) {
            buf[bufBase + i] = get(i);
        }
        return end;
    }

    /**
     * Copy a sequence of static arguments into a given buffer.
     * This is equivalent to {@code end-offset} separate calls to {@code get},
     * for each index in the range from {@code offset} up to but not including {@code end}.
     * Any static arguments that cannot be resolved are replaced by the
     * given sentinel value.
     * @param start index of first static argument to retrieve
     * @param end limiting index of static arguments to retrieve
     * @param buf array to receive the requested values
     * @param pos position in the array to offset storing the static arguments
     * @param ifNotPresent sentinel value to store if a static argument is not available
     * @return the limiting index, {@code end}
     * @throws LinkageError if {@code resolve} is true and a static argument cannot be resolved
     */
    default int copyConstants(int start, int end,
                              Object[] buf, int pos,
                              Object ifNotPresent) {
        int bufBase = pos - start;  // buf[bufBase + i] = get(i)
        for (int i = start; i < end; i++) {
            buf[bufBase + i] = get(i, ifNotPresent);
        }
        return end;
    }


    /** Returns the bootstrap method for this call.
     * @return the bootstrap method
     */
    MethodHandle bootstrapMethod();

    /** Returns the method name or constant name for this call.
     * @return the method name or constant name
     */
    String invocationName();

    /** Returns the method type or constant type for this call.
     * @return the method type or constant type
     */
    T invocationType();

    /**
     * Make a new bootstrap call descriptor with the given components.
     * @param bsm bootstrap method
     * @param name invocation name
     * @param type invocation type
     * @param args the additional static arguments for the bootstrap method
     * @param <T> the type of the invocation type, either {@link MethodHandle} or {@link Class}
     * @return a new bootstrap call descriptor with the given components
     */
    static <T> BootstrapCallInfo<T> makeBootstrapCallInfo(MethodHandle bsm,
                                                          String name,
                                                          T type,
                                                          Object... args) {
        AbstractConstantGroup.BSCIWithCache<T> bsci = new AbstractConstantGroup.BSCIWithCache<>(bsm, name, type, args.length);
        final Object NP = AbstractConstantGroup.BSCIWithCache.NOT_PRESENT;
        bsci.initializeCache(Arrays.asList(args), NP);
        return bsci;
    }

    /**
     * Invoke a bootstrap method handle with arguments obtained by resolving
     * the sequence of constants supplied by a given bootstrap call descriptor,
     * {@code bci}.
     * The first argument to the method will be {@code lookup}.
     * The second argument will be the invocation name of {@code bci}.
     * The third argument will be the invocation type of {@code bci}.
     * The fourth and subsequent arguments (if any) will be the resolved
     * constants, in order, supplied by {@code bci}.
     * <p>
     * @apiNote
     * This method behaves like the following but may be more optimal:
     * <blockquote><pre>{@code
     *   ArrayList<Object> args = new ArrayList<>();
     *   args.add(lookup);
     *   args.add(bsci.invocationName());
     *   args.add(bsci.invocationType());
     *   List<Object> constantArgs = bsci.asList();
     *   args.addAll(constantArgs);
     *   return handle.invokeWithArguments(args);
     * }</pre></blockquote>
     *
     * @param handle the bootstrap method handle to be invoked with resolved
     *        constants supplied by {@code bci}
     * @param lookup the lookup
     * @param bsci the bootstrap call descriptor
     * @return the result of invocation
     * @throws Throwable if an error occurs when resolving the constants from
     *         the bootstrap call descriptor or invoking the method handle
     */
    // @@@ More stuff to add as api note
    // This method is static so that it's possible to look it up and bind
    // to a method handle, thereby viewing that method handle as if accepts
    // a BootstrapCallInfo
    static Object invokeFromCallInfoToArguments(MethodHandle handle,
                                                MethodHandles.Lookup lookup,
                                                BootstrapCallInfo<?> bsci) throws Throwable {
        int argc = bsci.size();
        switch (argc) {
            case 0:
                return handle.invoke(lookup, bsci.invocationName(), bsci.invocationType());
            case 1:
                return handle.invoke(lookup, bsci.invocationName(), bsci.invocationType(),
                                     bsci.get(0));
            case 2:
                return handle.invoke(lookup, bsci.invocationName(), bsci.invocationType(),
                                     bsci.get(0), bsci.get(1));
            case 3:
                return handle.invoke(lookup, bsci.invocationName(), bsci.invocationType(),
                                     bsci.get(0), bsci.get(1), bsci.get(2));
            case 4:
                return handle.invoke(lookup, bsci.invocationName(), bsci.invocationType(),
                                     bsci.get(0), bsci.get(1), bsci.get(2), bsci.get(3));
            case 5:
                return handle.invoke(lookup, bsci.invocationName(), bsci.invocationType(),
                                     bsci.get(0), bsci.get(1), bsci.get(2), bsci.get(3), bsci.get(4));
            case 6:
                return handle.invoke(lookup, bsci.invocationName(), bsci.invocationType(),
                                     bsci.get(0), bsci.get(1), bsci.get(2), bsci.get(3), bsci.get(4), bsci.get(5));
            default:
                final int NON_SPREAD_ARG_COUNT = 3;  // (lookup, name, type)
                final int MAX_SAFE_SIZE = MethodType.MAX_MH_ARITY / 2 - NON_SPREAD_ARG_COUNT;
                if (argc >= MAX_SAFE_SIZE) {
                    // to be on the safe side, use invokeWithArguments which handles jumbo lists
                    Object[] newargv = new Object[NON_SPREAD_ARG_COUNT + argc];
                    newargv[0] = lookup;
                    newargv[1] = bsci.invocationName();
                    newargv[2] = bsci.invocationType();
                    bsci.copyArguments(0, argc, newargv, NON_SPREAD_ARG_COUNT);
                    return handle.invokeWithArguments(newargv);
                }
                MethodType invocationType = MethodType.genericMethodType(NON_SPREAD_ARG_COUNT + argc);
                MethodHandle typedBSM = handle.asType(invocationType);
                MethodHandle spreader = invocationType.invokers().spreadInvoker(NON_SPREAD_ARG_COUNT);
                Object[] argv = new Object[argc];
                bsci.copyArguments(0, argc, argv, 0);
                return spreader.invokeExact(typedBSM, (Object) lookup, (Object) bsci.invocationName(), bsci.invocationType(), argv);
        }
    }

    /**
     * Invoke a bootstrap method handle with a bootstrap call descriptor
     * argument composed from a given sequence of arguments.
     * <p>
     * @apiNote
     * This method behaves like the following but may be more optimal:
     * <blockquote><pre>{@code
     *   BootstrapCallInfo<Object> bsci = makeBootstrapCallInfo(handle, name, type, args);
     *   return handle.invoke(lookup, bsci);
     * }</pre></blockquote>
     *
     * @param handle the bootstrap method handle to be invoked with a bootstrap
     *        call descriptor composed from the sequence of arguments
     * @param lookup the lookup
     * @param name the method name or constant name
     * @param type the method type or constant type
     * @param args the sequence of arguments
     * @return the result of invocation
     * @throws Throwable if an error occurs when invoking the method handle
     */
    static Object invokeFromArgumentsToCallInfo(MethodHandle handle,
                                                MethodHandles.Lookup lookup,
                                                String name,
                                                Object type,
                                                Object... args) throws Throwable {
        BootstrapCallInfo<?> bsci = makeBootstrapCallInfo(handle, name, type, args);
        return handle.invoke(lookup, bsci);
    }
}