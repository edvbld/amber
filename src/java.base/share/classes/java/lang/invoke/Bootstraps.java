/*
 * Copyright (c) 2017, Oracle and/or its affiliates. All rights reserved.
 * DO NOT ALTER OR REMOVE COPYRIGHT NOTICES OR THIS FILE HEADER.
 *
 * This code is free software; you can redistribute it and/or modify it
 * under the terms of the GNU General Public License version 2 only, as
 * published by the Free Software Foundation.  Oracle designates this
 * particular file as subject to the "Classpath" exception as provided
 * by Oracle in the LICENSE file that accompanied this code.
 *
 * This code is distributed in the hope that it will be useful, but WITHOUT
 * ANY WARRANTY; without even the implied warranty of MERCHANTABILITY or
 * FITNESS FOR A PARTICULAR PURPOSE.  See the GNU General Public License
 * version 2 for more details (a copy is included in the LICENSE file that
 * accompanied this code).
 *
 * You should have received a copy of the GNU General Public License version
 * 2 along with this work; if not, write to the Free Software Foundation,
 * Inc., 51 Franklin St, Fifth Floor, Boston, MA 02110-1301 USA.
 *
 * Please contact Oracle, 500 Oracle Parkway, Redwood Shores, CA 94065 USA
 * or visit www.oracle.com if you need additional information or have any
 * questions.
 */
package java.lang.invoke;

import sun.invoke.util.Wrapper;

import java.util.AbstractMap;
import java.util.List;
import java.util.Map;

import static java.lang.invoke.MethodHandleNatives.mapLookupExceptionToError;
import static java.lang.invoke.MethodHandles.Lookup;

/**
 * Bootstrap methods for dynamically computed constants.
 */
public final class Bootstraps {
    // implements the upcall from the JVM, MethodHandleNatives.linkDynamicConstant:
    /*non-public*/
    static Object makeConstant(MethodHandle bootstrapMethod,
                               // Callee information:
                               String name, Class<?> type,
                               // Extra arguments for BSM, if any:
                               Object info,
                               // Caller information:
                               Class<?> callerClass) {
        // BSMI.invoke handles all type checking and exception translation.
        // If type is not a reference type, the JVM is expecting a boxed
        // version, and will manage unboxing on the other side.
        return BootstrapMethodInvoker.invoke(
                type, bootstrapMethod, name, type, info, callerClass);
    }

    /**
     * Returns the default value for a variable of the type specified in the
     * {@code type} argument, as per JLS 4.12.5 ({@code null} for reference types,
     * zero for primitive numeric types, {@code false} for boolean.)
     *
     * @param lookup unused
     * @param name unused
     * @param type the type for which we are seeking the default value
     * @param <T> the type for which we are seeking the default value (or the
     *           corresponding box type, if the type is a primitive type)
     * @return the default value
     */
    public static <T> T defaultValue(Lookup lookup, String name, Class<T> type) {
        return (type.isPrimitive())
               ? Wrapper.forPrimitiveType(type).zero(type)
               : null;
    }

    /**
     * Returns the {@link Class} mirror for a primitive type from its type
     * descriptor.
     *
     * @param lookup unused
     * @param name the descriptor (JVMS 4.3) of the desired primitive type
     * @param type the required result type (must be {@code Class.class})
     * @return the {@link Class} mirror
     * @throws IllegalArgumentException if the name is not a descriptor for a
     *         primitive type
     */
    @SuppressWarnings("rawtypes")
    public static Class<?> primitiveClass(Lookup lookup, String name, Class<Class> type) {
        if (name == null || name.length() == 0 || name.length() > 1)
            throw new IllegalArgumentException("not primitive: " + name);
        return Wrapper.forPrimitiveType(name.charAt(0)).primitiveType();
    }

    /**
     * Returns the value of a static final field.
     *
     * @param lookup the lookup context describing the class performing the
     *               operation (normally stacked by the JVM)
     * @param name the name of the field
     * @param type the type of the field
     * @param declaringClass the class in which the field is declared
     * @param <T> the type if the field (or the corresponding box type,
     *           if the type is a primitive type)
     * @return the value of the field
     * @throws IllegalAccessError if the declaring class or the field is not
     * accessible to the class performing the operation
     * @throws IncompatibleClassChangeError if the specified field is not {@code final}
     * @throws NoSuchFieldError if the specified field does not exist
     */
    public static <T> T getStaticFinal(Lookup lookup, String name, Class<T> type,
                                       Class<?> declaringClass) {
        MethodHandle mh;
        try {
            mh = lookup.findStaticGetter(declaringClass, name, type);
            MemberName member = mh.internalMemberName();
            if (!member.isFinal()) {
                throw new IncompatibleClassChangeError("not a final field: " + name);
            }
        }
        catch (ReflectiveOperationException ex) {
            throw mapLookupExceptionToError(ex);
        }

        try {
            // No need to cast because type was used to look up the MH
            @SuppressWarnings("unchecked")
            T value = (T) (Object) mh.invoke();
            return value;
        }
        catch (Error | RuntimeException e) {
            throw e;
        }
        catch (Throwable e) {
            throw new BootstrapMethodError(e);
        }
    }

    /**
     * Returns the value of a static final field whose declaring class is the
     * same as the field's type.  This is a simplified form of {@link Bootstraps#getStaticFinal(Lookup, String, Class, Class)}
     * for the case where a class declares distinguished constant instances of
     * itself.
     *
     * @param lookup the lookup context describing the class performing the
     *               operation (normally stacked by the JVM)
     * @param name the name of the field
     * @param type the type of the field
     * @param <T> the type if the field (or the corresponding box type,
     *           if the type is a primitive type)
     * @return the value of the field
     * @throws IllegalAccessError if the declaring class or the field is not
     * accessible to the class performing the operation
     * @throws IncompatibleClassChangeError if the specified field is not {@code final}
     * @throws NoSuchFieldError if the specified field does not exist
     * @see Bootstraps#getStaticFinal(Lookup, String, Class, Class)
     */
    public static <T> T getStaticFinal(Lookup lookup, String name, Class<T> type) {
        return getStaticFinal(lookup, name, type, promote(type));
    }

    static Class<?> promote(Class<?> type) {
        if (type.isPrimitive()) {
            type = Wrapper.forPrimitiveType(type).wrapperType();
        }
        return type;
    }


    /**
     * Returns the value of invoking a method handle with the provided arguments.
     *
     * @param lookup the lookup context describing the class performing the
     *               operation (normally stacked by the JVM)
     * @param name unused
     * @param type the type of the value to be returned, which must be compatible
     *             with the return type of the method handle
     * @param handle the method handle to be invoked
     * @param args the arguments to pass to the method handle, as if with
     * {@code invokeWithArguments}
     * @param <T> the type of the value to be returned (or the corresponding box
     *           type, if the type is a primitive type)
     * @return the result of invoking the method handle
     * @throws WrongMethodTypeException if the type is incompatible with the
     *         handle's return type
     * @throws Throwable TODO
     */
    public static <T> T invoke(Lookup lookup, String name, Class<T> type,
                               MethodHandle handle, Object... args) {
        if (type != handle.type().returnType()) {
            // Convert if the method handle return type and the constanr type
            // differ
            handle.asType(handle.type().changeReturnType(type));
        }
        try {
            @SuppressWarnings("unchecked")
            T t = (T) handle.invokeWithArguments(args);
            return t;
        }
        catch (Error | RuntimeException e) {
            throw e;
        }
        catch (Throwable e) {
            throw new BootstrapMethodError(e);
        }
    }

    /**
     * Finds a {@link VarHandle} for an instance field.
     *
     * @param lookup the lookup context describing the class performing the
     *               operation (normally stacked by the JVM)
     * @param name the name of the field
     * @param type unused; must be {@code Class<VarHandle>}
     * @param decl the type declaring the field
     * @param fieldType the field type
     * @return the {@code VarHandle}
     * @throws LinkageError if the {@code VarHandle} cannot be found
     * @throws NullPointerException if any unused argument is {@code null}
     */
    public static VarHandle varHandleInstanceField(MethodHandles.Lookup lookup, String name, Class<VarHandle> type,
                                                   Class<?> decl, Class<?> fieldType) {
        try {
            return lookup.findVarHandle(decl, name, fieldType);
        }
        catch (ReflectiveOperationException e) {
            throw mapLookupExceptionToError(e);
        }
    }

    /**
     * Finds a {@link VarHandle} for a static field.
     *
     * @param lookup the lookup context describing the class performing the
     *               operation (normally stacked by the JVM)
     * @param name the name of the field
     * @param type unused; must be {@code Class<VarHandle>}
     * @param decl the type declaring the field
     * @param fieldType the field type
     * @return the handle
     * @throws LinkageError if the {@code VarHandle} cannot be found
     * @throws NullPointerException if any unused argument is {@code null}
     */
    public static VarHandle varHandleStaticField(MethodHandles.Lookup lookup, String name, Class<VarHandle> type,
                                                 Class<?> decl, Class<?> fieldType) {
        try {
            return lookup.findStaticVarHandle(decl, name, fieldType);
        }
        catch (ReflectiveOperationException e) {
            throw mapLookupExceptionToError(e);
        }
    }

    /**
     * Finds a {@link VarHandle} for an array type.
     *
     * @param lookup unused; the lookup context describing the class performing
     *               the operation (normally stacked by the JVM)
     * @param name unused
     * @param type unused; must be {@code Class<VarHandle>}
     * @param arrayClass the array type
     * @return the handle
     * @throws IllegalArgumentException if arrayClass is not an array type
     * @throws NullPointerException if any unused argument is {@code null}
     */
    public static VarHandle varHandleArray(MethodHandles.Lookup lookup, String name, Class<VarHandle> type,
                                           Class<?> arrayClass) {
        return MethodHandles.arrayElementVarHandle(arrayClass);
    }

    /**
     * Return an unmodifiable list containing the specified elements.
     *
     * @param lookup unused
     * @param name unused
     * @param type unused, must be {@code Class<List<?>>}
     * @param elements the elements of the list
     * @return the list
     */
    public static List<?> list(Lookup lookup, String name, Class<List<?>> type, Object... elements) {
        return List.of(elements);
    }

    /**
     * Return an unmodifiable map containing the specified mappings.
     *
     * @param lookup unused
     * @param name unused
     * @param type unused, must be {@code Class<Map<?, ?>>}
     * @param elements the keys and values of the map, alternating
     * @return the map
     * @throws IllegalArgumentException if an odd number of elements is provided
     */
    public static Map<?, ?> map(Lookup lookup, String name, Class<Map<?, ?>> type, Object... elements) {
        if (elements.length % 2 != 0) {
            throw new IllegalArgumentException("Odd number of keys and values");
        } else {
            Map.Entry<?,?>[] entries = new Map.Entry<?, ?>[elements.length / 2];
            for (int i=0; i<elements.length / 2; i++)
                entries[i] = new AbstractMap.SimpleImmutableEntry<>(elements[2*i], elements[2*i+1]);
            return Map.ofEntries(entries);
        }
    }
}
