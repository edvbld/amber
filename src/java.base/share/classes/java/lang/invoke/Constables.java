/*
 * Copyright (c) 2017, Oracle and/or its affiliates. All rights reserved.
 * DO NOT ALTER OR REMOVE COPYRIGHT NOTICES OR THIS FILE HEADER.
 *
 * This code is free software; you can redistribute it and/or modify it
 * under the terms of the GNU General Public License version 2 only, as
 * published by the Free Software Foundation.  Oracle designates this
 * particular file as subject to the "Classpath" exception as provided
 * by Oracle in the LICENSE file that accompanied this code.
 *
 * This code is distributed in the hope that it will be useful, but WITHOUT
 * ANY WARRANTY; without even the implied warranty of MERCHANTABILITY or
 * FITNESS FOR A PARTICULAR PURPOSE.  See the GNU General Public License
 * version 2 for more details (a copy is included in the LICENSE file that
 * accompanied this code).
 *
 * You should have received a copy of the GNU General Public License version
 * 2 along with this work; if not, write to the Free Software Foundation,
 * Inc., 51 Franklin St, Fifth Floor, Boston, MA 02110-1301 USA.
 *
 * Please contact Oracle, 500 Oracle Parkway, Redwood Shores, CA 94065 USA
 * or visit www.oracle.com if you need additional information or have any
 * questions.
 */

package java.lang.invoke;

import java.util.stream.Stream;

/**
 * Classes for representing entries in the constant pool, which can be
 * intrinsified via methods in {@link Intrinsics}.
 */
public class Constables {
    static final ClassRef CLASS_CONDY = ClassRef.of("java.lang.invoke.Bootstraps");

    static final MethodHandleRef BSM_GET_STATIC_FINAL_SELF
            = MethodHandleRef.ofCondyBootstrap(CLASS_CONDY, "getStaticFinal", ClassRef.CR_Object);
    static final MethodHandleRef BSM_GET_STATIC_FINAL_DECL
            = MethodHandleRef.ofCondyBootstrap(CLASS_CONDY, "getStaticFinal", ClassRef.CR_Object, ClassRef.CR_Class);
    static final MethodHandleRef BSM_DEFAULT_VALUE
            = MethodHandleRef.ofCondyBootstrap(CLASS_CONDY, "defaultValue", ClassRef.CR_Object);
    static final MethodHandleRef BSM_VARHANDLE
            = MethodHandleRef.ofCondyBootstrap(CLASS_CONDY, "varHandle", ClassRef.CR_VarHandle, ClassRef.CR_MethodType, ClassRef.CR_Object.array());

    static final ConstantRef<?> NULL = ConstantRef.ofNull();

    /**
     * Resolve a {@link ConstantRef} relative to the provided lookup.  The {@link ConstantRef}
     * must be one of the following types: {@link String}, {@link Integer},
     * {@link Long}, {@link Float}, {@link Double}, {@link ClassRef},
     * {@link MethodTypeRef}, {@link MethodHandleRef}, or {@link DynamicConstantRef}.
     *
     * @param c The {@link ConstantRef}
     * @param lookup The lookup object to use
     * @param <T> The type of the object described by the {@link ConstantRef}
     * @return The resolved object
     * @throws ReflectiveOperationException If there is an error resolving the
     * constable
     */
    @SuppressWarnings("unchecked")
    static<T> T resolve(ConstantRef<T> c, MethodHandles.Lookup lookup) throws ReflectiveOperationException {
        if (c instanceof String
            || (c instanceof Integer)
            || (c instanceof Long)
            || (c instanceof Float)
            || (c instanceof Double))
            return (T) c;
        if (c instanceof ClassRef) {
            return (T) ((ClassRef) c).resolve(lookup);
        }
        else if (c instanceof MethodTypeRef) {
            return (T) ((MethodTypeRef) c).resolve(lookup);
        }
        else if (c instanceof MethodHandleRef) {
            return (T) ((MethodHandleRef) c).resolve(lookup);
        }
        else if (c instanceof DynamicConstantRef) {
            return (T) ((DynamicConstantRef) c).resolve(lookup);
        }
        else
            throw new IllegalArgumentException(c.getClass().getName());
    }

    static Object[] resolveArgs(MethodHandles.Lookup lookup, ConstantRef<?>[] args) {
        return Stream.of(args)
                     .map(arg -> {
                         try {
                             return resolve(arg, lookup);
                         }
                         catch (ReflectiveOperationException e) {
                             throw new RuntimeException(e);
                         }
                     })
                     .toArray();
    }

    /**
     * Returns a {@link ConstantRef}, if the argument is not a {@link ClassRef}
     * then the argument is returned unchanged. If the argument is a {@link ClassRef}
     * and it is a primitive {@link ClassRef}, then a fresh {@link DynamicConstantRef}
     * corresponding to it is created and returned.
     *
     * @param <T> The type of the object described by the {@link ConstantRef}
     * @param ref the given {@link ConstantRef}
     * @return the reduced {@link ConstantRef}
     */
    public static<T> ConstantRef<T> reduce(ConstantRef<T> ref) {
        if (ref instanceof ClassRef) {
            ClassRef cr = (ClassRef) ref;
            if (cr.isPrimitive()) {
                // Return a dynamic constant whose value is obtained by getting
                // static final TYPE field on the boxed class
                return DynamicConstantRef.of(BootstrapSpecifier.of(BSM_GET_STATIC_FINAL_DECL, cr.promote()),
                                             "TYPE", ClassRef.CR_Class);
            }
        }
        return ref;
    }

    /**
     * Convert a class literal to a descriptor string
     * @param clazz the class literal
     * @return the descriptor string
     */
    static String classToDescriptor(Class<?> clazz) {
        return MethodType.methodType(clazz).toMethodDescriptorString().substring(2);
    }
}