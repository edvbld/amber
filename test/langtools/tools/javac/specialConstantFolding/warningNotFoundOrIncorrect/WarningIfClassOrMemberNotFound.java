/*
 * @test /nodynamiccopyright/
 * @summary warn if a class or member is not found at compile time
 * @compile/fail/ref=WarningIfClassOrMemberNotFound.out -Werror -XDrawDiagnostics WarningIfClassOrMemberNotFound.java
 */

import java.lang.invoke.*; import java.lang.constant.*;
import static java.lang.invoke.Intrinsics.*;

class WarningIfClassOrMemberNotFound {
    void m() {
        final MethodTypeDesc mt1 = MethodTypeDesc.of(ClassDesc.ofDescriptor("Ljava/lang/String;"));
        MethodHandle mh1 = ldc(MethodHandleDesc.of(DirectMethodHandleDesc.Kind.STATIC, ClassDesc.ofDescriptor("LNonExistentClass1;"), "foo", mt1));

        ClassDesc cr = ClassDesc.ofDescriptor("LNonExistentClass2;");
        Class<?> c = ldc(cr);

        ClassDesc crArr = ClassDesc.ofDescriptor("[[[LNonExistentClass3;");
        Class<?> cArr = ldc(crArr);

        final MethodTypeDesc mt2 = MethodTypeDesc.of(ClassDesc.ofDescriptor("Ljava/lang/String;"));
        // now the class exists but the method doesn't
        MethodHandle mh2 = ldc(MethodHandleDesc.of(DirectMethodHandleDesc.Kind.STATIC, ClassDesc.ofDescriptor("LWarningIfClassOrMemberNotFound;"), "bar", mt2));

        final MethodTypeDesc mt3 = MethodTypeDesc.of(ClassDesc.ofDescriptor("Ljava/lang/String;"));
        // now the class exists and so is the method but the arguments are incorrect
        MethodHandle mh3 = ldc(MethodHandleDesc.of(DirectMethodHandleDesc.Kind.STATIC, ClassDesc.ofDescriptor("LWarningIfClassOrMemberNotFound;"), "foo", mt3));
    }

    String foo(String s) { return s; }
}
