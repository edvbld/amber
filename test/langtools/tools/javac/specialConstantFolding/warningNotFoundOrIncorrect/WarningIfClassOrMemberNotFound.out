WarningIfClassOrMemberNotFound.java:13:31: compiler.warn.class.not.found: NonExistentClass1
WarningIfClassOrMemberNotFound.java:16:25: compiler.warn.class.not.found: NonExistentClass2
WarningIfClassOrMemberNotFound.java:19:28: compiler.warn.class.not.found: NonExistentClass3
WarningIfClassOrMemberNotFound.java:23:31: compiler.warn.member.not.found.at.class: bar, WarningIfClassOrMemberNotFound
WarningIfClassOrMemberNotFound.java:27:31: compiler.warn.member.not.found.at.class: foo, WarningIfClassOrMemberNotFound
- compiler.err.warnings.and.werror
1 error
5 warnings
